pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarClassLoader.java

```diff
@@ -135,21 +143,34 @@ public class NarClassLoader extends URLClassLoader {
      * The NAR for which this <tt>ClassLoader</tt> is responsible.
      */
     private final File narWorkingDirectory;
+    private final Pair<String, ClassLoader> classLoaderCacheKey;
+    private final AtomicInteger refCnt = new AtomicInteger(1);
 
     private static final String TMP_DIR_PREFIX = "pulsar-nar";
 
     public static final String DEFAULT_NAR_EXTRACTION_DIR = System.getProperty("nar.extraction.tmpdir") != null
             ? System.getProperty("nar.extraction.tmpdir") : System.getProperty("java.io.tmpdir");
 
-    static NarClassLoader getFromArchive(File narPath, Set<String> additionalJars, ClassLoader parent,
-                                                String narExtractionDirectory)
-        throws IOException {
-        File unpacked = NarUnpacker.unpackNar(narPath, getNarExtractionDirectory(narExtractionDirectory));
-        return AccessController.doPrivileged(new PrivilegedAction<NarClassLoader>() {
-            @SneakyThrows
-            @Override
-            public NarClassLoader run() {
-                return new NarClassLoader(unpacked, additionalJars, parent);
+    static NarClassLoader getFromArchive(File narFile, Set<String> additionalJars, ClassLoader parent,
+                                         String narExtractionDirectory) throws IOException {
+
+        String md5Sum = Base64.getUrlEncoder().withoutPadding().encodeToString(NarUnpacker.calculateMd5sum(narFile));
```

Comment:
`calculateMd5sum` is a private method, which cannot be called here.

And could you add a test for it?